File: cpp/021-030/Reverse Nodes in k-Group.cpp
class Solution {
    
    int length(ListNode* head) {
        int len = 0;
        while (head != nullptr) {
            head = head->next;
            len++;
        }
        return len;
    }
    
    ListNode* dfs(ListNode* head, int len, int k) {
        if (len < k) {
            return head;
        }
        ListNode* tail = head;
        
        ListNode* prev = nullptr, *tmp = nullptr;
        for (int i = 0; i < k; i++) {
            tmp = head->next;
            head -> next = prev;
            prev = head;
            head = tmp;
        }
        
        tail->next = dfs(head, len - k, k);
        return prev;
    }
    
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        return dfs(head, length(head), k);
    }
};

File: python/21-30/Reverse Nodes in k-Group.py
#!/usr/bin/env python
# coding=utf-8

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param {ListNode} head
    # @param {integer} k
    # @return {ListNode}
    def reverseKGroup(self, head, k):
        if not head or k < 2:
            return head

        ret = head
        for i in range(k-1):
            ret = ret.next
            if not ret:
                return head

        prev,cur = None,head
        for i in range(k):
            tmp = cur.next
            cur.next = prev
            prev = cur
            cur = tmp

        head.next = self.reverseKGroup(cur, k)
        return ret